### DISCUSSION TITLE
General inquiries

### DISCUSSION THREAD

**Discussion by thorsten-gehrig** [rednblkx/HomeKey-ESP32#2] [General]
Hi Rednblkx
i´m very interested into your HomeKey implementation.
I´m running myself since some years and NFC reader based on Arduino and connected to KNX Home-automation bus... always with the wish to upgrade to HomeKey.
Reading your code I´m asking myself for some key topics:
a) how to connect the ESP to a network? I dont see any wifi code there... - any guidance?
b) how to enroll a phone/where to get the pairing code? (I´m able to run the Apple-Home-Key-Reader python code - those does show the code on the console)
c) is there any external communication planned - e.g. MQTT? 
last but not least:
is there anything i can help? I´m not a bit coder - but i may can help adding code for WiFi, MQTT, KNX due to libraries... (but I´m not promising anything... I´m more more an home-automation enthusiastic)

Thank you so far for this GREAT project!
Kind regards
Thorsten

**Comment by juaurodri** [comment]
A Esphome external component would be perfect 👌
Thanks for your work on this esp implementation

**Reply by rednblkx** [reply]
at first i was going to make an esphome component, however the lack of a HAP component made me rather choose the present approach but i do want to revisit that sometime in the future.

**Comment by rednblkx** [comment]
Hi Thorsten,

Thank you for taking interest in this project.

First of all the project was built upon [HomeSpan](https://github.com/HomeSpan/HomeSpan) which provides the HomeKit integration and handles everything, which enabled me to focus on the HomeKey implementation.

So to answer your questions:
1. You'll need to use the serial terminal in order to connect it to wifi, instructions will be displayed upon booting, HomeSpan does implement a temp wifi but it involves a physical button to enter configuration mode that i don't have wired up since i already use the terminal.
2. The HomeKit pairing code is the default 466-37-726 from HomeSpan as seen in their [documentation](https://github.com/HomeSpan/HomeSpan/blob/master/docs/UserGuide.md#pairing-to-homekit) that you can change at runtime if you wish.
3. Plan is to initially implement MQTT and have it expose events so you can have a device trigger in Home Assistant and i actually had a basic MQTT implementation during my tests..... so it's coming

As to how you can help, i'll take whatever suggestions you have, however on the code side at the moment i'm working on refactoring the code to tighten up some loose ends and make it more organized.

_**This discussion will remain open for any suggestions or inquiries anyone may have.**_

**Reply by thorsten-gehrig** [reply]
I´m amazed!
it does work :-) :-) GREAT
It does also detect other NFC cards ("Found something" followed by the card-serial) - that is GREAT
(and i need to have those also exposed via MQTT).
I´m super-motivated -> several new ESP and NFC-readers are ordered via Aliexpress :-) 
(just to have enough spare to tinker)
I´m looking forward for MQTT implementation (guessing there are ready-to-use libraries) - more than happy to help testing.

As you did ask for suggestions: currently the authentication does trigger unlock -> next one lock -> next one unlock ...
For several doors they just open for something like 3-5 seconds and then automatically lock if the door wasn't opened.
For those use cases it might make sense to implement an automatic status change to locked after a defined time (3-5 seconds).

Kind regards
Thorsten

PS: I´m still amazed 👍 and full of endorphins

**Comment by Klayman42** [comment]
Hi rednblkx,

first of all: Thanks for the great work!! After being totally disappointed from the new ekey devices, this opens up completely new possibilities. I was able to successfully set up your project and connect it to a motorized lock controlled by a KNX actuator. I have two questions on how to continue:

- Is it possible to use the same home key across multiple locks, like front and back entrance or garage door? What hapens if add a second instance of your homekey implementation to my home?
- How much effort would it be to run this on an Ethernet based and PoE powered ESP32 like the ESP32-POE from Olimex? Or would that interfere with the HomeSpan library and its WiFi functionality?

Kind regards,
Klayman

**Reply by kormax** [reply]
I can answer the first question:

- You can add as many locks as needed. In this case HomeKit will configure all of the other locks with the same cryptographic keys and identifiers that were generated during the provisioning of the first one, so the same "pass" can be used for all of them.

- The "pass" on a device keeps up to 16 last used persistent keys (which are unique to each lock/device combo) in memory for FAST auth (<1s), but even if you manage to add and use more, It'll fallback to STANDARD (~2s) or ATTESTATION auth (~4s).

**Reply by rednblkx** [reply]
Thank you @kormax

I have to mention that i haven't yet implemented ATTESTATION but it's in the works, it was previously but i had to remove it during restructuring.

Regarding the second question, i haven't tested this but i don't think Ethernet would work since HomeSpan relies heavily on the WIFI component however i'm looking to maybe migrate to espressif's [homekit sdk](https://github.com/espressif/esp-homekit-sdk) or something else sometime in the future so i could some more in-depth and customised control and Ethernet will be included.

For short, @Klayman42
- Apple designed the Home Key so you have one key(pass) per Apple Home that can be used with multiple locks within the home
- Ethernet is not usable currently as HomeSpan relies on WIFI
